Add fet custom pattern

LuighiV opened the pull request:
Add possibility to create any patter with an array of fets and do an automatic routing.

Based on custom patterning implemented for #50 

Note: If required we can remove commits from BJT just to enclose for custom pattern

LuighiV committed:
Add custom patterning to fet

LuighiV committed:
Add notebook to test custom patterns with fet

Nimok15 reviewed (changes requested):
line 623 of [469b72e], We are taking the base geometry for the new emitter port. This could cause layer/port mismatch.

LuighiV commented:
Hi @Nimok15  
I think I've added the requested change, can you take a look on it please?
As this PR is built on PR #50 I've added the correction to both. 

Thanks,
Luighi

Nimok15 commented:
Hi @LuighiV, thanks for updating it. I will review #50 and get back to you.  

Thanks,
Nimish

LuighiV committed:
Fix generation when there is just 1 row

LuighiV committed:
Add gate_shared and src_shared options for the routing

LuighiV committed:
Add tests for shared gate and source

LuighiV committed:
Add comments
